milla/src/lib.rs

```diff
@@ -0,0 +1,2224 @@
+                        tile_atmos[ATMOS_SUPERCONDUCTIVITY_EAST] = OPEN_HEAT_TRANSFER_COEFFICIENT;
+                        tile_atmos[ATMOS_SUPERCONDUCTIVITY_SOUTH] = OPEN_HEAT_TRANSFER_COEFFICIENT;
+                        tile_atmos[ATMOS_SUPERCONDUCTIVITY_WEST] = OPEN_HEAT_TRANSFER_COEFFICIENT;
+                    }
+                }
+            }
+            $buffer_map.insert($z, atmos_lock);
+        }
+    };
+}
+
+// Fetches the buffer out of the buffer map, creating it if need be, and strips it down to just the
+// boxed array.
+macro_rules! get_atmos {
+    ( $var: ident, $buffer_map: ident, $z: expr ) => {
+        init_atmos!($buffer_map, $z);
+        // Sometimes we need this mutable, sometimes not, so allow it to be mut
+        // even if it's unneeded.
+        #[allow(unused_mut)]
+        let mut $var = $buffer_map.get(&$z).unwrap().write().unwrap();
+    };
+}
+
+// Fetches the buffer out of the buffer map, without initializing or stripping off the RwLock.
+// Prefer gat_atmos if you can use it.
+macro_rules! get_atmos_raw {
+    ( $var: ident, $buffer: ident, $z: expr ) => {
+        let $var = $buffer.get(&$z).unwrap();
+    };
+}
+
+// Extracts a small array for a given tile out of the buffer's big array.
+macro_rules! get_tile {
+    ( $var: ident, $atmos: expr, $x: expr, $y: expr ) => {
+        let $var: &[f32; ATMOS_DEPTH] = (&$atmos
+            [calc_index($x, $y, 0)..calc_index($x, $y, ATMOS_DEPTH)])
+            .try_into()
+            .unwrap();
+    };
+}
+
+// Extracts a small mutable array for a given tile out of the buffer's big array.
+macro_rules! get_mutable_tile {
+    ( $var: ident, $atmos: expr, $x: expr, $y: expr ) => {
+        let $var: &mut [f32; ATMOS_DEPTH] = (&mut $atmos
+            [calc_index($x, $y, 0)..calc_index($x, $y, ATMOS_DEPTH)])
+            .try_into()
+            .unwrap();
+    };
+}
+
+// Returns an Option<[f32; ATMOS_DEPTH]> that is the get_tile result for the tile, if it's valid,
+// or None otherwise.
+macro_rules! maybe_get_tile {
+    ( $atmos: expr, $x: expr, $y: expr ) => {
+        if $x < 0 || $x >= MAP_SIZE as i32 || $y < 0 || $y >= MAP_SIZE as i32 {
+            None
+        } else {
+            Some(
+                (&$atmos[calc_index($x, $y, 0)..calc_index($x, $y, ATMOS_DEPTH)])
+                    .try_into()
+                    .unwrap(),
+            )
+        }
+    };
+}
+
+// Calculates the index of the specified position in any buffer's array.
+fn calc_index(x: i32, y: i32, offset: usize) -> usize {
+    usize::try_from(x).unwrap() * MAP_SIZE * ATMOS_DEPTH
+        + usize::try_from(y).unwrap() * ATMOS_DEPTH
+        + offset
+}
+
+// Turns a BYOND number into an Option<f32>.
+// The option will be None if the number was null or NaN.
+fn byond_to_option_f32(value: ByondValue) -> Result<Option<f32>, Error> {
+    if value.is_null() {
+        Ok(None)
+    } else {
+        Ok(f32_to_option_f32(value.get_number()?))
+    }
```

Comment:
I think this clears?
```suggestion
    Ok(if value.is_null() {None} else {f32_to_option_f32(value.get_number()?)})
```

Reply:
Whether it works or not, I find that version *significantly* harder to read.

Reply:
Valid!